bridges/snowbridge/primitives/router/src/outbound/v2.rs

```diff
@@ -0,0 +1,1704 @@
+			return Err(SendError::NotApplicable)
+		}
+
+		let source_location = Location::new(1, local_sub.clone());
+
+		let agent_id = match AgentHashedDescription::convert_location(&source_location) {
+			Some(id) => id,
+			None => {
+				log::error!(target: TARGET, "unroutable due to not being able to create agent id. '{source_location:?}'");
+				return Err(SendError::NotApplicable)
+			},
+		};
+
+		let message = message.clone().ok_or_else(|| {
+			log::error!(target: TARGET, "xcm message not provided.");
+			SendError::MissingArgument
+		})?;
+
+		// Inspect AliasOrigin as V2 message
+		let mut instructions = message.clone().0;
+		let result = instructions.matcher().match_next_inst_while(
+			|_| true,
+			|inst| {
+				return match inst {
+					AliasOrigin(..) => Err(ProcessMessageError::Yield),
+					_ => Ok(ControlFlow::Continue(())),
+				}
+			},
+		);
+		ensure!(result.is_err(), SendError::NotApplicable);
+
+		let mut converter =
+			XcmConverter::<ConvertAssetId, ()>::new(&message, expected_network, agent_id);
+		let message = converter.convert().map_err(|err| {
+			log::error!(target: TARGET, "unroutable due to pattern matching error '{err:?}'.");
+			SendError::Unroutable
+		})?;
+
+		// validate the message
+		let (ticket, _) = OutboundQueue::validate(&message).map_err(|err| {
+			log::error!(target: TARGET, "OutboundQueue validation of message failed. {err:?}");
+			SendError::Unroutable
+		})?;
+
+		Ok(((ticket.encode(), XcmHash::from(message.id)), Assets::default()))
+	}
+
+	fn deliver(blob: (Vec<u8>, XcmHash)) -> Result<XcmHash, SendError> {
+		let ticket: OutboundQueue::Ticket = OutboundQueue::Ticket::decode(&mut blob.0.as_ref())
+			.map_err(|_| {
+				log::trace!(target: TARGET, "undeliverable due to decoding error");
+				SendError::NotApplicable
+			})?;
+
+		let message_id = OutboundQueue::deliver(ticket).map_err(|_| {
+			log::error!(target: TARGET, "OutboundQueue submit of message failed");
+			SendError::Transport("other transport error")
+		})?;
+
+		log::info!(target: TARGET, "message delivered {message_id:#?}.");
+		Ok(message_id.into())
+	}
+}
+
+/// Errors that can be thrown to the pattern matching step.
+#[derive(PartialEq, Debug)]
+pub enum XcmConverterError {
+	UnexpectedEndOfXcm,
+	EndOfXcmMessageExpected,
+	WithdrawAssetExpected,
+	DepositAssetExpected,
+	NoReserveAssets,
+	FilterDoesNotConsumeAllAssets,
+	TooManyAssets,
+	ZeroAssetTransfer,
+	BeneficiaryResolutionFailed,
+	AssetResolutionFailed,
+	InvalidFeeAsset,
+	SetTopicExpected,
+	ReserveAssetDepositedExpected,
+	InvalidAsset,
+	UnexpectedInstruction,
+	TooManyCommands,
+	AliasOriginExpected,
+	InvalidOrigin,
+}
+
+macro_rules! match_expression {
+	($expression:expr, $(|)? $( $pattern:pat_param )|+ $( if $guard: expr )?, $value:expr $(,)?) => {
+		match $expression {
+			$( $pattern )|+ $( if $guard )? => Some($value),
+			_ => None,
+		}
+	};
+}
+
+pub struct XcmConverter<'a, ConvertAssetId, Call> {
```

Comment:
There is going to be a lot more conversion code for v2, so I suggest moving `XcmConverter` to a new file like `outbound/v2/converter.rs`.